lib/ansible/module_utils/basic.py

```diff
@@ -427,6 +432,88 @@ def remove_values(value, no_log_strings):
     return new_value
 
 
+def sanitize_keys(obj, no_log_strings, ignore_keys=None):
+    """ Sanitize the keys in a container object by removing no_log values from key names.
+
+    This is a companion function to the `remove_values()` function. Similar to that function,
+    we make use of deferred_removals to avoid hitting maximum recursion depth in cases of
+    large data structures.
+
+    :param obj: The container object to sanitize. Non-container objects are returned unmodified.
+    :param no_log_strings: A set of string values we do not want logged.
+    :param ignore_keys: A set of string values of keys to not sanitize.
+
+    :returns: An object with sanitized keys.
+    """
+
+    if ignore_keys is None:
+        ignore_keys = frozenset([])
+
+    deferred_removals = deque()
+
+    no_log_strings = [to_native(s, errors='surrogate_or_strict') for s in no_log_strings]
+    new_value = _sanitize_keys(obj, no_log_strings, ignore_keys, deferred_removals)
+
+    while deferred_removals:
+        old_data, new_data = deferred_removals.popleft()
+
+        if isinstance(new_data, Mapping):
+            for old_key, old_elem in old_data.items():
+                if old_key not in ignore_keys and not old_key.startswith('_ansible'):
+                    # Sanitize the old key. We take advantage of the sanitizing code in
+                    # _remove_values_conditions() rather than recreating it here. Since the
+                    # value should always be a string-like object, it's safe to pass None
+                    # for the deferred_removals parameter.
+                    new_key = _remove_values_conditions(old_key, no_log_strings, None)
+                    new_data[new_key] = _sanitize_keys(old_elem, no_log_strings, ignore_keys, deferred_removals)
+                else:
+                    new_data[old_key] = _sanitize_keys(old_elem, no_log_strings, ignore_keys, deferred_removals)
+        else:
+            for elem in old_data:
+                new_elem = _sanitize_keys(elem, no_log_strings, ignore_keys, deferred_removals)
+                if isinstance(new_data, MutableSequence):
+                    new_data.append(new_elem)
+                elif isinstance(new_data, MutableSet):
+                    new_data.add(new_elem)
+                else:
+                    raise TypeError('Unknown container type encountered when removing private values from keys')
+
+    return new_value
+
+
+def _sanitize_keys(value, no_log_strings, ignore_keys, deferred_removals):
```

Comment:
Note: We put callers below callees in the majority of ansible files, including this one.  So `_sanitize_keys()` should be above `sanitize_keys()` in the file.

Reply:
Also note... perhaps we should use a similar name to the` _remove_values` helper (so change this to `_sanitize_keys_conditions`) ?

Reply:
Hmmm.... Now that I've read the code more thoroughly..... This function doesn't sanitize the keys at all ;-)

It's more like `_find_containers_to_process`...

Reply:
I renamed it anyway. I don't want to bikeshed on it.